File: 2nd semester [Erasmus+]/PECS (Performance Evaluation of Computer Systems)/task-01-timers/02-timers/timers.py
from timeit import timeit
from time import perf_counter


def sample(count):
    print("Sampling range({}) using timeit...".format(count))
    time = timeit(lambda: range(count))
    print("Sampled range({}) using timeit.".format(count))
    return time


def sample2(count):
    print("Sampling range({}) using perf_counter...".format(count))
    start = perf_counter()
    range(count)
    stop = perf_counter()
    time = stop - start
    print("Sampled range({}) using perf_counter.".format(count))
    return time


def write_file(filename, header, counts, times, times2):
    print("Writing file {}...".format(filename))
    with open(filename, "w") as out_file:
        out_file.write("{}\n".format(header))
        for count, time, time2 in zip(counts, times, times2):
            out_file.write("{},{},{}\n".format(count, time, time2))
    print("Wrote file {}.".format(filename))


def worker():
    counts = [100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000, 10000000, 50000000, 100000000]
    times = [sample(count) for count in counts]
    times2 = [sample2(count) for count in counts]

    write_file("timers_python.csv", "count,timeit,perf_counter", counts, times, times2)


if __name__ == '__main__':
    worker()

File: 2nd semester [Erasmus+]/PECS (Performance Evaluation of Computer Systems)/task-01-timers/02-timers/Timers.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Timers {

    public static void main(String[] args) {
        worker();
    }

    private static void initializeArray(int size) {
        int[] array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = 0;
        }
    }

    private static void writeFile(String filename, String header, List<Integer> sizes, List<Long> times, List<Long> times2) {
        System.out.printf("Writing file %s ...%n", filename);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            writer.write(String.format("%s\n", header));
            for (int i = 0; i < sizes.size(); i++) {
                writer.write(String.format("%d,%d,%d\n", sizes.get(i), times.get(i), times2.get(i)));
            }
        } catch (IOException ioException) {
            System.out.printf("Error writing file %s .%n", filename);
        }
        System.out.printf("Written file %s .%n", filename);
    }

    public static void worker() {
        List<Integer> sizes = List.of(100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000, 10000000, 50000000);
        List<Long> times = new ArrayList<>();
        List<Long> times2 = new ArrayList<>();
        long start, stop, time;

        for (int size : sizes) {
            System.out.printf("Measuring array initialization of size %s...%n", size);

            start = System.currentTimeMillis();
            initializeArray(size);
            stop = System.currentTimeMillis();
            time = stop - start;
            times.add(time);

            start = System.nanoTime();
            initializeArray(size);
            stop = System.nanoTime();
            time = stop - start;
            times2.add(time);

            System.out.printf("Measured array initialization of size %s.%n", size);
        }

        writeFile("timers_java.csv", "size,currentTimeMillis,nanoTime", sizes, times, times2);
    }
}
